hand_tracking_publisher.cpp

```diff
@@ -0,0 +1,296 @@
+                    _current_depth = cv::Mat();  // must calculate from parallax
+                } else {
+                    // TODO there is an issue
+                }
+            } else {
+                _current_depth = rgb_depth_ptr->at(data_format::image::DEPTH);
+            }
+        } else {
+            _current_depth = depth_ptr->at(data_format::image::DEPTH);
+        }
+    }
+
+    if (_current_raw.confidence_valid_) {
+        _current_confidence = _current_raw.at(data_format::image::CONFIDENCE);
+    } else {
+        _current_confidence = cv::Mat();
+    }
+    Eigen::Matrix3f rot;
+    if (_current_pose.valid)
+        rot =_current_pose.orientation.toRotationMatrix();
+
+    for (auto& item : _detections)
+        data_format::denormalize(item.second, static_cast<float>(_img_size_x), static_cast<float>(_img_size_y), data_format::units::PIXEL);
+    // only use left eye detections, as the depth map is expressed as left eye distance
+    for (auto h : data_format::ht::hand_map) {
+        auto& primary_eye = _detections.at(_current_raw.primary).points.at(h);
+        if (!primary_eye.valid) {
+            continue;
+        }
+
+        data_format::ht::hand_points hand_pnts(_current_pose.unit);
+        for (int i = 0 ; i < data_format::ht::NUM_LANDMARKS; i++) {
+            if (primary_eye[i].x() == 0. || primary_eye[i].y() == 0. || !primary_eye[i].valid) {
+                hand_pnts[i].confidence = 0.;
+                hand_pnts[i].valid = false;
+            } else {
+                Eigen::Vector3f pnt;
+
+                ///// need to adjust angle if it is to the left or right of center
+                auto primary = cam_data_[_current_raw.primary];
+                const int primary_x = static_cast<int>(primary_eye[i].x());
+                const int primary_y = static_cast<int>(primary_eye[i].y());
+                double theta_xl = std::atan((primary.center_x - primary_eye[i].x()) * std::tan(primary.horizontal_fov / 2.) / primary.center_x);
+                double theta_yl = std::atan((primary.center_y - primary_eye[i].y()) * std::tan(primary.vertical_fov / 2.) / primary.center_y);
+                float distance = 0;
+                double confidence = -1.;
+                if (!_current_confidence.empty())
+                    confidence = 1. - (_current_confidence.at<float>(primary_y, primary_x)) / 100.;
+                if (!_current_depth.empty())
+                    distance = std::abs(_current_depth.at<float>(primary_y, primary_x));
+
+                // use parallax to determine distance
+                if ((!_current_confidence.empty() && confidence <= .05) ||
+                    distance <= data_format::units::convert(_current_pose.unit, data_format::units::MILLIMETER, 10)
+                    || distance >= data_format::units::convert(_current_pose.unit, data_format::units::METER, 20.)) {
+                    if (_current_raw.eye_count == 1) {
+                        // TODO emit warning
```

Comment:
Remaining TODO. How can we even use parallax if there's only one eye? This should throw an error.